File: 204-count-primes/count-primes.cpp
class Solution {
public:
    int countPrimes(int n) {

        // Initialize a count variable to keep track of the number of prime numbers found.
        int count = 0;

        // Vector of boolean values to keep track of prime numbers. 
        // Initially, all numbers are assumed to be prime (true), except for indices 0 and 1.
        vector<bool> check (n+1, true);
        check[0] = check[1] = false; // 0 and 1 are not prime numbers.

        // Loop through all numbers from 2 to n to check for primality.
        for(int i = 2; i < n; i++){
            // If the current number is marked as prime,
            // increment the prime count and mark its multiples as non-prime.
            if(check[i]){
                count++;
                // Mark all multiples of i as non-prime. Starting from i*i because
                // any number less than i*i would have already been marked by smaller primes.
                for(long long j = (long long)i*i; j<n; j+=i){
                    check[j] = false;
                }
            }
        }
        // Return the total count of prime numbers found.
        return count;
    }
};

File: 204-count-primes/count-primes.py
class Solution:
    def countPrimes(self, n: int) -> int:
        if n < 2:
            return 0
        
        # Initialize a list to mark non-prime numbers
        is_not_prime = [0] * n
        
        # Iterate from 2 to n-1 to mark non-prime numbers
        for i in range(2, n):
            if is_not_prime[i] == 0:  # If i is a prime number
                # Mark multiples of i starting from i*i as non-prime
                j = i * i
                while j < n:
                    is_not_prime[j] = 1
                    j += i
        
        # Count prime numbers by counting 0s in the is_not_prime list
        # prime_count = 0
        # for i in range(2, len(is_not_prime)):
        #     if is_not_prime[i] == 0:
        #         prime_count += 1
        return n - 2 - sum(is_not_prime)
        # return prime_count
